src/Spectre.libGaussianMixtureModelling/ExpectationMaximization.h

```diff
@@ -0,0 +1,160 @@
+                std::mt19937_64& rngEngine, const unsigned numberOfComponents = 2)
+            : m_pMzArray(mzArray), m_pIntensities(intensities), m_DataSize(size), m_Components(numberOfComponents)
+            , m_Initialization(mzArray, size, m_Components, rngEngine)
+            , m_Expectation(mzArray, size, m_AffilationMatrix, m_Components)
+            , m_Maximization(mzArray, intensities, size, m_AffilationMatrix, m_Components)
+            , m_LogLikelihoodCalculator(mzArray, intensities, size, m_Components)
+        {
+            if (mzArray == nullptr)
+            {
+                throw ArgumentNullException("mzArray");
+            }
+
+            if (intensities == nullptr)
+            {
+                throw ArgumentNullException("intensities");
+            }
+
+            m_AffilationMatrix = new DataType*[size];
+            if (size)
+            {
+                m_AffilationMatrix[0] = new DataType[numberOfComponents * size];
+                for (unsigned i = 1; i < size; ++i)
+                    m_AffilationMatrix[i] = m_AffilationMatrix[0] + i * numberOfComponents;
+            }
+        }
+
+        /// <summary>
+        /// Performs a full algorithm run. Terminates when change in log likelihood
+        /// is sufficiently small (lower than  0.00000001).
+        /// </summary>
+        /// <returns>
+        /// Gaussian Mixture Model containing all the components with their approprite
+        /// parameters.
+        /// </returns>
+        GaussianMixtureModel ExpectationMaximization::EstimateGmm()
+        {
+            Initialization();
+
+            DataType oldLikelihood; // used as iterations terminator
+            DataType newLikelihood = CalculateLikelihood();
+            do
+            {
+                Expectation();
+                oldLikelihood = newLikelihood;
+                Maximization();
+                newLikelihood = CalculateLikelihood();
+            } while (abs(oldLikelihood - newLikelihood) > 0.00000001);
+
+            return GaussianMixtureModel(
+                gsl::span<DataType>(m_pMzArray, m_DataSize),
+                gsl::span<DataType>(m_pIntensities, m_DataSize),
+                std::move(m_Components)
+            );
+        }
+
+        /// <summary>
+        /// Calculates the log likelihood using the class supplied template type.
+        /// </summary>
+        /// <returns>
+        /// Value of log likelihood.
+        /// </returns>
+        DataType ExpectationMaximization::CalculateLikelihood()
+        {
+            return m_LogLikelihoodCalculator.CalculateLikelihood();
+        }
+
+        /// <summary>
+        /// Destructor deleting the affilation matrix, also known as probability 
+        /// matrix
+        /// </summary>
+        ExpectationMaximization::~ExpectationMaximization()
+        {
+            if (m_DataSize) delete[] m_AffilationMatrix[0];
+            delete[] m_AffilationMatrix;
```

Comment:
Yes, because it is the one that that stores the data used throughout the whole array.

The initialization works that way:
- Dynamically allocate array of DataType pointers
- Allocate <size_of_the_matrix> bytes and assign it to the first pointer
- Assign the rest of the pointers to appropriate parts of the memory allocated to the first pointer